src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,144 @@
+  set.maps_to T U U ↔ U.invariant_under T := iff.rfl
+
+/-- `U` is `T` invariant is equivalent to saying `T(U) ⊆ U` -/
+lemma submodule.invariant_under_iff (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ T '' U ⊆ U := by rw [← set.maps_to', U.invariant_under_iff_maps_to]
+
+variables (U V : submodule R E) (hUV : is_compl U V) (T : E →ₗ[R] E)
+
+local notation `eᵤ` := submodule.linear_proj_of_is_compl U V hUV
+local notation `eᵥ` := submodule.linear_proj_of_is_compl V U (is_compl.symm hUV)
+
+lemma submodule.proj_comp_self_comp_proj_eq_of_invariant_under
+  (h : U.invariant_under T) (x : E) : ↑(eᵤ (T ↑(eᵤ x))) = T ↑(eᵤ x) :=
+begin
+  rw submodule.linear_proj_of_is_compl_eq_self_iff,
+  exact h (submodule.coe_mem _),
+end
+
+lemma submodule.invariant_under_of_proj_comp_self_comp_proj_eq
+  (h : ∀ x : E, ↑(eᵤ (T ↑(eᵤ x))) = T ↑(eᵤ x)) : U.invariant_under T :=
+begin
+  intros u hu,
+  rw [submodule.mem_comap, ← submodule.linear_proj_of_is_compl_eq_self_iff hUV _,
+      ← (submodule.linear_proj_of_is_compl_eq_self_iff hUV u).mpr hu, h],
+end
+
+lemma submodule.proj_comp_self_comp_proj_eq_iff_invariant_under :
+  U.invariant_under T ↔ (∀ x : E, ↑(eᵤ (T ↑(eᵤ x))) = T ↑(eᵤ x)) :=
+begin
+ split,
+ { intros h x,
+   exact submodule.proj_comp_self_comp_proj_eq_of_invariant_under U V hUV T h x, },
+ { intros h,
+   exact submodule.invariant_under_of_proj_comp_self_comp_proj_eq U V hUV T h, }
+end
+
+lemma submodule.compl_invariant_under_iff_linear_proj_and_T_commute :
+  (U.invariant_under T ∧ V.invariant_under T) ↔ commute (U.subtype.comp (eᵤ)) T :=
+begin
+  rw [commute, semiconj_by, linear_map.ext_iff],
+  simp only [linear_map.mul_apply, linear_map.coe_comp, submodule.coe_subtype,
+             function.comp_app, submodule.proj_comp_self_comp_proj_eq_iff_invariant_under U V hUV,
+             submodule.proj_comp_self_comp_proj_eq_iff_invariant_under V U (is_compl.symm hUV)],
+  have : ∀ x : E, ↑(eᵥ x) = x - ↑(eᵤ x) := λ x,
+  by rw [eq_sub_iff_add_eq, add_comm, submodule.linear_proj_add_linear_proj_of_is_compl_eq_self],
+  simp_rw [this], clear this,
+  simp_rw [map_sub, sub_eq_self, add_subgroup_class.coe_sub,
+           sub_eq_zero, ← submodule.proj_comp_self_comp_proj_eq_iff_invariant_under],
+  split,
+  { rintros ⟨h1, h2⟩ x,
+    simp_rw [h2 x],
+    exact (submodule.linear_proj_of_is_compl_eq_self_iff hUV _).mpr
+      (h1 (set_like.coe_mem (eᵤ x))), },
+  { intros h,
+    split,
+    { intros u h',
+      specialize h u,
+      simp_rw [(submodule.linear_proj_of_is_compl_eq_self_iff hUV _).mpr h'] at h,
+      exact (submodule.linear_proj_of_is_compl_eq_self_iff hUV _).mp h, },
+    { intros x,
+      simp_rw [← h, submodule.linear_proj_of_is_compl_apply_left], } },
+end
+
+lemma submodule.commutes_with_linear_proj_iff_linear_proj_eq [invertible T] :
+  commute (U.subtype.comp eᵤ) T ↔
+    (⅟ T).comp ((U.subtype.comp eᵤ).comp T) = U.subtype.comp eᵤ :=
```

Comment:
My feeling would be to avoid all the `invertible` stuff by assuming `T` to be `linear_equiv` here, but maybe that will backfire later in applications. @eric-wieser what do you think ?